In [1]:
import numpy as np
import pandas as pd 

### READING DATA USING PANDAS

We use pandas read.csv function to read the csv file in python and pandas DataFrame method to convert file into data frame.

In [3]:
# Reading data using Pandas

Titanic = pd.DataFrame(pd.read_csv('/content/train (1).csv'))
print("Titanic has shape:",Titanic.shape) #returns no.of rows, columns

Titanic has shape: (891, 12)


we have loaded the train(1).csv file into a Pandas DataFrame and it consists of 891 rows and 12 columns. Now let's look at the Titanic data


In [5]:
Titanic

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0              1         0       3  ...   7.2500   NaN         S
1              2         1       1  ...  71.2833   C85         C
2              3         1       3  ...   7.9250   NaN         S
3              4         1       1  ...  53.1000  C123         S
4              5         0       3  ...   8.0500   NaN         S
..           ...       ...     ...  ...      ...   ...       ...
886          887         0       2  ...  13.0000   NaN         S
887          888         1       1  ...  30.0000   B42         S
888          889         0       3  ...  23.4500   NaN         S
889          890         1       1  ...  30.0000  C148         C
890          891         0       3  ...   7.7500   NaN         Q

[891 rows x 12 columns]

We can take a look at the first 5 rows of data using the .head() method, as shown below.

In [6]:
Titanic.head()

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S
3            4         1       1  ...  53.1000  C123         S
4            5         0       3  ...   8.0500   NaN         S

[5 rows x 12 columns]

We can also take a look at the last 5 rows of data by using the .tail() method, as shown below

In [7]:
Titanic.tail()

PassengerId  Survived  Pclass  ...   Fare Cabin  Embarked
886          887         0       2  ...  13.00   NaN         S
887          888         1       1  ...  30.00   B42         S
888          889         0       3  ...  23.45   NaN         S
889          890         1       1  ...  30.00  C148         C
890          891         0       3  ...   7.75   NaN         Q

[5 rows x 12 columns]

### HANDLING NULL VALUES

Let's do a quick check to see whether we have any null values in our dataset. To do this, we will use the .isnull() method followed by the .any() method to check whether any of the columns contain null values.

In [8]:
Titanic.isnull().any()

PassengerId    False
Survived       False
Pclass         False
Name           False
Sex            False
Age             True
SibSp          False
Parch          False
Ticket         False
Fare           False
Cabin           True
Embarked        True
dtype: bool

In Pandas, logical True values have numerical value 1 and logical False values have numerical value 0. Therefore, we can count the number of null values by counting the number of logical True values.

In [9]:
# We count the number of null values in the columns of Titanic
Titanic.isnull().sum()  # returns a pandas series with columns name as the label index

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

Hence, we can see that this dataset contains columns with null values. In such cases we have two options:

 - Either drop the entire row or column
 - Fill the missing values with some appropriate value. Like mean of all the values in that column. 

Separating out the columns which have more than 35% of the values missing in the Dataset 

In [26]:
#cabin has not much importance to our data. So we can drop it.
x = Titanic.isnull().sum()  
drop_col = x[x>(40/100 * df.shape[0])]  # we are storing only those columns 
                                        #which have more than 40% of the data missing. 
drop_col


Series([], dtype: int64)

In [27]:
drop_col.index

Index([], dtype='object')

In [28]:
Titanic.drop(drop_col.index, axis = 1, inplace = True )  #axis = 1 for columns 
Titanic.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Embarked         2
dtype: int64

In [30]:
# For age, we can fill with mean of column
Titanic.fillna(Titanic.mean(),inplace = True) #all the null values are replaced by mean of column
Titanic.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       2
dtype: int64

still Embarked contains null value. Because, this column contains strings and strings doesn't have mean.

In [31]:
Titanic['Embarked'].describe() #returns all the details of Embarked column
# count tells the no.of non-null values. i.e, 891-2 = 889
# unique returns no.of unique values. i.e, s,c,q
# top returns element with highest frequency.(i.e, most occuring element)
# freq returns no.of times that top element returns. i.e, S occurs 644 times here.

count     889
unique      3
top         S
freq      644
Name: Embarked, dtype: object

For Embarked attribute, we fill the null values with the most frequent element in the column

In [32]:
Titanic['Embarked'].fillna('S', inplace =True)
Titanic.isnull().sum() #Now all the null values have been filled

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64

In [36]:
# corr() returns correlation between each columns
#propotanality between values of 2 columns
Titanic.corr()

PassengerId  Survived    Pclass  ...     SibSp     Parch      Fare
PassengerId     1.000000 -0.005007 -0.035144  ... -0.057527 -0.001652  0.012658
Survived       -0.005007  1.000000 -0.338481  ... -0.035322  0.081629  0.257307
Pclass         -0.035144 -0.338481  1.000000  ...  0.083081  0.018443 -0.549500
Age             0.033207 -0.069809 -0.331339  ... -0.232625 -0.179191  0.091566
SibSp          -0.057527 -0.035322  0.083081  ...  1.000000  0.414838  0.159651
Parch          -0.001652  0.081629  0.018443  ...  0.414838  1.000000  0.216225
Fare            0.012658  0.257307 -0.549500  ...  0.159651  0.216225  1.000000

[7 rows x 7 columns]

Here, the corr between Fare column and Pclass col is good, which tells the Fare is increasing with Pclass.
For Pclass1 Fare is higher and for Pclass3 Fare is lesser(negative correlation)


We can make a new column family_size by combaining 2 columns.

- sibsp:Number of siblings/spouses Abroad

- parch:Number of Parents/Children Abroad

In [38]:
Titanic['Family_size'] = Titanic['SibSp'] + Titanic['Parch'] #new column with sum of both values
Titanic.drop(['SibSp','Parch'], axis=1, inplace = True)
Titanic.corr()


PassengerId  Survived    Pclass       Age      Fare  Family_size
PassengerId     1.000000 -0.005007 -0.035144  0.033207  0.012658    -0.040143
Survived       -0.005007  1.000000 -0.338481 -0.069809  0.257307     0.016639
Pclass         -0.035144 -0.338481  1.000000 -0.331339 -0.549500     0.065997
Age             0.033207 -0.069809 -0.331339  1.000000  0.091566    -0.248512
Fare            0.012658  0.257307 -0.549500  0.091566  1.000000     0.217138
Family_size    -0.040143  0.016639  0.065997 -0.248512  0.217138     1.000000

- Family_size in the ship does not have much correlance with survival rate(0.016639).
- Family_size have more correlation with fare. Hence as family size increases fare increases.
- Survived has major corr with Pclass. So, for Pclass1 survival rate will be more and for pclass3 survival rate will be less (negative correaltion).


Lets check if the person was alone or not van affect the survival rate.

In [39]:
Titanic['Alone'] = [0 if Titanic['Family_size'][i]>0 else 1 for i in Titanic.index]
#when family_size is 0 , person is alone.So value 1 for alone. if family_size >0 not alone(alone = 0).
Titanic.head()

PassengerId  Survived  Pclass  ... Embarked Family_size  Alone
0            1         0       3  ...        S           1      0
1            2         1       1  ...        C           1      0
2            3         1       3  ...        S           0      1
3            4         1       1  ...        S           1      0
4            5         0       3  ...        S           0      1

[5 rows x 11 columns]

In [45]:
Titanic.groupby(['Alone'])['Survived'].mean()
#calculate mean of survived corresponding to alone = 0, 
#calculate mean of survived corresponding to alone = 1

Alone
0    0.505650
1    0.303538
Name: Survived, dtype: float64

If the person is alone, he has less chance of surviving.
- The reason might be the person who is traveling with his family might be belonging to rich class.

In [46]:
Titanic[['Alone','Fare']].corr()

Alone      Fare
Alone  1.000000 -0.271832
Fare  -0.271832  1.000000

Here it is negative corr. So we can say if the person was not alone, then chance for the ticket price is higher

In [47]:
Titanic['Sex'] = [0 if Titanic["Sex"][i] == 'male' else 1 for i in Titanic.index] #1 for female, 0 for male
Titanic.groupby(['Sex'])["Survived"].mean()

Sex
0    0.188908
1    0.742038
Name: Survived, dtype: float64

It shows female pssengers have more chance of surviving than male ones.

In [48]:
Titanic.groupby(['Embarked'])["Survived"].mean()

Embarked
C    0.553571
Q    0.389610
S    0.339009
Name: Survived, dtype: float64

It shows percentage of survival is more for the persons who has embarkation port as C wheras minimum for embarkation port as S.

###**CONCLUSION**

- People in high class have higher survival rate than lowerclass people.
- As family size increases fare also increases.
- If the person is alone, he has less chance of surviving. The reason might be the person who is traveling with his family might be belonging to rich class.
- Female passengers have more chance of surviving than male ones. Female passengers were prioritized over man.
- Survivalrate is more for the persons who has embarkation port as Cherbourg.
